File: mobile/lib/models/ride.dart
class Ride {
  final String id;
  final String sessionId;
  final String bikeId;
  final double totalDistanceKm;
  final int durationSeconds;
  final int movingDurationSeconds;
  final double averageSpeedKmh;
  final double maxSpeedKmh;
  final double elevationGainM;
  final double elevationLossM;
  final String? startAddress;
  final String? endAddress;
  final String? encodedPolyline;
  final DateTime startedAt;
  final DateTime endedAt;

  Ride({
    required this.id,
    required this.sessionId,
    required this.bikeId,
    required this.totalDistanceKm,
    required this.durationSeconds,
    required this.movingDurationSeconds,
    required this.averageSpeedKmh,
    required this.maxSpeedKmh,
    required this.elevationGainM,
    required this.elevationLossM,
    this.startAddress,
    this.endAddress,
    this.encodedPolyline,
    required this.startedAt,
    required this.endedAt,
  });

  factory Ride.fromJson(Map<String, dynamic> json) {
    return Ride(
      id: json['id'] ?? '',
      sessionId: json['session_id'] ?? '',
      bikeId: json['bike_id'] ?? '',
      totalDistanceKm: (json['total_distance_km'] as num?)?.toDouble() ?? 0.0,
      durationSeconds: json['duration_seconds'] ?? 0,
      movingDurationSeconds: json['moving_duration_seconds'] ?? 0,
      averageSpeedKmh: (json['average_speed_kmh'] as num?)?.toDouble() ?? 0.0,
      maxSpeedKmh: (json['max_speed_kmh'] as num?)?.toDouble() ?? 0.0,
      elevationGainM: (json['elevation_gain_m'] as num?)?.toDouble() ?? 0.0,
      elevationLossM: (json['elevation_loss_m'] as num?)?.toDouble() ?? 0.0,
      startAddress: json['start_address'],
      endAddress: json['end_address'],
      encodedPolyline: json['encoded_polyline'],
      startedAt: DateTime.parse(json['started_at']),
      endedAt: DateTime.parse(json['ended_at']),
    );
  }
}

File: backend/app/models/ride.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, String, Float, Integer, DateTime, ForeignKey, Text, JSON
from sqlalchemy.orm import relationship
from app.core.database import Base


class Ride(Base):
    __tablename__ = "rides"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id = Column(String(36), ForeignKey("tracking_sessions.id", ondelete="CASCADE"), unique=True, nullable=False)
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    bike_id = Column(String(36), ForeignKey("bikes.id", ondelete="CASCADE"), nullable=False, index=True)
    
    # Ride Core Metrics
    total_distance_km = Column(Float, default=0.0, nullable=False)
    duration_seconds = Column(Integer, default=0, nullable=False)
    moving_duration_seconds = Column(Integer, default=0, nullable=False)
    average_speed_kmh = Column(Float, default=0.0, nullable=False)
    max_speed_kmh = Column(Float, default=0.0, nullable=False)
    elevation_gain_m = Column(Float, default=0.0, nullable=False)
    elevation_loss_m = Column(Float, default=0.0, nullable=False)
    
    # Locations
    start_latitude = Column(Float, nullable=True)
    start_longitude = Column(Float, nullable=True)
    start_address = Column(String(255), nullable=True)
    end_latitude = Column(Float, nullable=True)
    end_longitude = Column(Float, nullable=True)
    end_address = Column(String(255), nullable=True)
    
    # Polyline & GeoJSON for fast map rendering
    encoded_polyline = Column(Text, nullable=True)
    route_geojson = Column(JSON, nullable=True)
    
    # Timestamps
    started_at = Column(DateTime(timezone=True), nullable=False, index=True)
    ended_at = Column(DateTime(timezone=True), nullable=False, index=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)

    # Relationships
    session = relationship("TrackingSession", back_populates="ride")
    bike = relationship("Bike", back_populates="rides")
